src/ripple/app/tx/impl/InvariantCheck.cpp

```diff
@@ -358,6 +358,101 @@ AccountRootsNotDeleted::finalize(
 
 //------------------------------------------------------------------------------
 
+void
+AccountRootsDeletedClean::visitEntry(
+    bool isDelete,
+    std::shared_ptr<SLE const> const& before,
+    std::shared_ptr<SLE const> const&)
+{
+    if (isDelete && before && before->getType() == ltACCOUNT_ROOT)
+        accountsDeleted_.emplace_back(before);
+}
+
+bool
+AccountRootsDeletedClean::finalize(
+    STTx const& tx,
+    TER const result,
+    XRPAmount const,
+    ReadView const& view,
+    beast::Journal const& j)
+{
+    // This list should include all of the keylet functions that take a single
+    // AccountID parameter, except nftpage_max, because it is exercised below
+    // using the NFT page lookup logic. (nftpage_min is included because it
+    // should _never_ be present.)
+    using fType = std::function<Keylet(AccountID const& id)>;
+    static std::array<fType, 4> const keyletfuncs{
+        &keylet::account,
+        &keylet::ownerDir,
+        &keylet::signers,
+        &keylet::nftpage_min,
+        //&keylet::nftpage_max
+    };
+
+    // Always check for objects in the ledger, but to prevent differing
+    // transaction processing results, however unlikely, only fail if the
+    // feature is enabled. Enabled, or not, though, a fatal-level message will
+    // be logged
+    bool const enforce = view.rules().enabled(featureInvariantsV1_1);
+
+    auto const objectExists = [&view, enforce, &j](auto const& keylet) {
+        if (auto const sle = view.read(keylet))
+        {
+            // Finding the object is bad
+            auto const typeName = [&sle]() {
+                auto item =
+                    LedgerFormats::getInstance().findByType(sle->getType());
+
+                if (item != nullptr)
+                    return item->getName();
+                return std::to_string(sle->getType());
+            }();
+
+            JLOG(j.fatal())
+                << "Invariant failed: account deletion left behind a "
+                << typeName << " object";
+            assert(enforce);
+            return true;
+        }
+        return false;
+    };
+
+    for (auto const& accountSLE : accountsDeleted_)
+    {
+        auto const accountID = accountSLE->getAccountID(sfAccount);
+        // Simple types
+        for (auto const& keyletfunc : keyletfuncs)
+        {
+            if (objectExists(std::invoke(keyletfunc, accountID)) && enforce)
+                return false;
+        }
+
+        // NFT pages
+        {
+            Keylet const first = keylet::nftpage_min(accountID);
+            Keylet const last = keylet::nftpage_max(accountID);
+
+            uint256 key =
+                view.succ(first.key, last.key.next()).value_or(last.key);
+
+            // current page
+            if (objectExists(Keylet{ltNFTOKEN_PAGE, key}) && enforce)
+                return false;
+        }
+
+        // Keys directly stored in the AccountRoot object
+        if (auto const ammKey = accountSLE->at(~sfAMMID))
```

Comment:
I think this is redundant since you are checking for the `ownerDir`. If AMM's owner directory is deleted then the AMM's object is deleted as well. One can't exist without the other.